File: scripts/ui/ContainerUI.cs
using Godot;
using ProiectDeLegume.Scripts.Inventory;
using ProiectDeLegume.Scripts.Localization;
using ProiectDeLegume.Scripts.World;

namespace ProiectDeLegume.Scripts.UI;

public partial class ContainerUI : PanelContainer
{
    public bool IsOpen { get; private set; }

    private ContainerData _currentContainer;
    private Inventory.Inventory _inventory;
    private Player.PlayerStats _stats;
    private VBoxContainer _itemList;
    private Label _titleLabel;
    private System.Action _onChanged;

    public override void _Ready()
    {
        CustomMinimumSize = new Vector2(320, 300);
        Size = new Vector2(320, 300);
        Position = new Vector2(20, 20);
        Visible = false;

        // Dark background
        var style = new StyleBoxFlat();
        style.BgColor = new Color(0.08f, 0.08f, 0.12f, 0.86f);
        style.BorderWidthBottom = style.BorderWidthTop = style.BorderWidthLeft = style.BorderWidthRight = 1;
        style.BorderColor = new Color(0.3f, 0.3f, 0.35f);
        AddThemeStyleboxOverride("panel", style);

        var vbox = new VBoxContainer();
        AddChild(vbox);

        _titleLabel = new Label();
        _titleLabel.AddThemeFontSizeOverride("font_size", 18);
        vbox.AddChild(_titleLabel);

        var sep = new HSeparator();
        vbox.AddChild(sep);

        _itemList = new VBoxContainer();
        vbox.AddChild(_itemList);

        var spacer = new Control { SizeFlagsVertical = SizeFlags.ExpandFill };
        vbox.AddChild(spacer);

        var closeHint = new Label();
        closeHint.Text = Lang.Get("container.close");
        closeHint.AddThemeFontSizeOverride("font_size", 12);
        closeHint.AddThemeColorOverride("font_color", new Color(0.4f, 0.4f, 0.4f));
        vbox.AddChild(closeHint);
    }

    public void Open(ContainerData container, Inventory.Inventory inventory, Player.PlayerStats stats, System.Action onChanged)
    {
        _currentContainer = container;
        _inventory = inventory;
        _stats = stats;
        _onChanged = onChanged;
        IsOpen = true;
        Visible = true;

        _titleLabel.Text = container.DisplayName;
        RefreshItems();
    }

    public void Close()
    {
        IsOpen = false;
        Visible = false;
        _currentContainer = null;
    }

    private void RefreshItems()
    {
        // Clear old
        foreach (var child in _itemList.GetChildren())
            child.QueueFree();

        var items = _currentContainer.Items;
        if (items.Count == 0)
        {
            var empty = new Label();
            empty.Text = Lang.Get("container.empty");
            empty.AddThemeColorOverride("font_color", new Color(0.4f, 0.4f, 0.4f));
            _itemList.AddChild(empty);
            return;
        }

        for (int i = 0; i < items.Count; i++)
        {
            var stack = items[i];
            var row = new HBoxContainer();

            var nameLabel = new Label();
            string countStr = stack.Count > 1 ? $" x{stack.Count}" : "";
            nameLabel.Text = $"{stack.Def.LocalizedName}{countStr}";
            nameLabel.SizeFlagsHorizontal = SizeFlags.ExpandFill;
            row.AddChild(nameLabel);

            var takeBtn = new Button();
            takeBtn.Text = Lang.Get("container.take");
            takeBtn.CustomMinimumSize = new Vector2(60, 0);
            int capturedIdx = i;
            takeBtn.Pressed += () =>
            {
                if (_inventory.CanAdd(stack.Def))
                {
                    _inventory.Add(stack.Def);
                    if (stack.Count > 1) stack.Count--;
                    else _currentContainer.Items.Remove(stack);
                    RefreshItems();
                    _onChanged?.Invoke();
                }
            };
            row.AddChild(takeBtn);

            var weightLabel = new Label();
            weightLabel.Text = $"{stack.TotalWeight:F1}{Lang.Get("unit.kg")}";
            weightLabel.AddThemeColorOverride("font_color", new Color(0.7f, 0.7f, 0.7f));
            row.AddChild(weightLabel);

            _itemList.AddChild(row);
        }
    }
}

File: src/commonMain/kotlin/ui/ContainerUI.kt
package ui

import inventory.Inventory
import inventory.ItemStack
import korlibs.image.color.*
import korlibs.korge.input.*
import korlibs.korge.view.*
import localization.Lang
import world.ContainerManager

class ContainerUI(private val root: Container) {
    private var panel: Container? = null
    private var currentContainer: ContainerManager.ContainerData? = null
    var isOpen: Boolean = false
        private set

    fun open(container: ContainerManager.ContainerData, items: List<ItemStack>, inventory: Inventory, onChanged: () -> Unit) {
        close()
        currentContainer = container
        isOpen = true

        panel = root.container {
            val panelW = 320.0
            val panelH = 300.0
            val px = 20.0
            val py = 20.0

            solidRect(panelW, panelH, RGBA(20, 20, 30, 220)).position(px, py)

            text(container.displayName, textSize = 18.0).position(px + 10, py + 8)

            solidRect(panelW - 20, 1.0, RGBA(100, 100, 100, 200)).position(px + 10, py + 32)

            if (items.isEmpty()) {
                text(Lang["container.empty"], textSize = 14.0, color = Colors["#666666"]).position(px + 10, py + 42)
            } else {
                for ((index, stack) in items.withIndex()) {
                    val yOff = py + 40 + index * 28.0

                    solidRect(panelW - 20, 24.0, RGBA(40, 40, 55, 200)).position(px + 10, yOff)

                    val countStr = if (stack.count > 1) " x${stack.count}" else ""
                    text("${stack.def.localizedName}$countStr", textSize = 13.0)
                        .position(px + 14, yOff + 4)

                    text("${formatWeight(stack.totalWeight)}${Lang["unit.kg"]}", textSize = 11.0, color = Colors["#cccccc"])
                        .position(px + panelW - 60, yOff + 5)

                    val btn = solidRect(44.0, 20.0, RGBA(60, 120, 60, 220))
                    btn.position(px + panelW - 115, yOff + 2)
                    text(Lang["container.take"], textSize = 10.0).position(px + panelW - 112, yOff + 6)

                    btn.onClick {
                        if (inventory.canAdd(stack.def)) {
                            inventory.add(stack.def)
                            if (stack.count > 1) {
                                stack.count--
                            } else {
                                container.items.remove(stack)
                            }
                            open(container, container.items, inventory, onChanged)
                            onChanged()
                        }
                    }
                }
            }

            text(Lang["container.close"], textSize = 12.0, color = Colors["#666666"])
                .position(px + 10, py + panelH - 22)
        }
    }

    fun close() {
        panel?.removeFromParent()
        panel = null
        currentContainer = null
        isOpen = false
    }
}

private fun formatWeight(w: Double): String {
    val rounded = (w * 10).toInt() / 10.0
    return if (rounded == rounded.toInt().toDouble()) "${rounded.toInt()}.0" else "$rounded"
}
